modules/auxiliary/scanner/rsync/modules_list.rb

```diff
@@ -6,64 +6,223 @@
+  def rsync_parse_lines(response_data)
+    control_lines = []
+    data_lines = []
+
+    if response_data
+      response_data.strip!
+      response_data.split(/\n/).map do |line|
+        if line =~ /^#{RSYNC_HEADER}/
+          control_lines << line
+        else
+          data_lines << line
+        end
+      end
+    end
+
+    [ control_lines, data_lines ]
   end
 
   def run_host(ip)
-    connect
-    version = sock.get_once
-
-    return if version.blank?
-
-    print_good("#{ip}:#{rport} - rsync #{version.strip} found")
-    report_service(:host => ip, :port => rport, :proto => 'tcp', :name => 'rsync')
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_version',
-        :data => version.strip
-    )
+    begin
+      connect
+      version, motd = rsync_negotiate
+      unless version
+        vprint_error("#{peer} - does not appear to be rsync")
+        disconnect
+        return
+      end
+    rescue => e
+      vprint_error("#{peer} - error while connecting and negotiating: #{e}")
+      disconnect
+      return
+    end
 
-    # making sure we match the version of the server
-    sock.puts("#{version}")
-    # the listing command
-    sock.puts("\n")
-    listing = sock.get(20)
-    disconnect
-
-    return if listing.blank?
-
-    print_good("#{ip}:#{rport} - rsync listing found")
-    listing.gsub!('@RSYNCD: EXIT', '') # not interested in EXIT message
-    listing_sanitized = Rex::Text.to_hex_ascii(listing.strip)
-
-    vprint_status("#{ip}:#{rport} - #{version.rstrip} #{listing_sanitized}")
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_listing',
-        :data => listing_sanitized
+    info = "rsync protocol version #{version}"
+    info += ", MOTD '#{motd}'" if motd
+    report_service(
+      host: ip,
+      port: rport,
+      proto: 'tcp',
+      name: 'rsync',
+      info: info
     )
+    print_status("#{peer} - rsync version: #{version}") if datastore['SHOW_VERSION']
+    print_status("#{peer} - rsync MOTD: #{motd}") if motd && datastore['SHOW_MOTD']
+
+    modules_metadata = {}
+    begin
+      modules_metadata = rsync_list
+    rescue => e
+      vprint_error("#{peer} -- error while listing modules: #{e}")
+      return
+    ensure
+      disconnect
+    end
+
+    if modules_metadata.empty?
+      print_status("#{peer} - no rsync modules found")
+    else
+      modules = modules_metadata.map { |m| m[:name] }
+      print_good("#{peer} - #{modules.size} rsync modules found: #{modules.join(', ')}")
+
+      table_columns = %w(Name Comment)
+      if datastore['TEST_AUTHENTICATION']
+        table_columns << 'Authentication'
+        modules_metadata.each do |module_metadata|
+          begin
+            connect
+            rsync_negotiate
+            module_metadata[:authentication] = get_rsync_auth_status(module_metadata[:name])
+          rescue => e
```

Comment:
Did you add this rescue to make sure the ensure kick in? Without the rescue it will kick in and call #disconnect anyway.

I ask because it's generally not a good practice to rescue all. You could be rescuing something that's not meant to be rescued by the module.... like for example, something msfconsole should handle. I can't think of anything specific right now, but this generally should be avoided. My suggestion is either you rescue specific exceptions, or you don't need to rescue anything at all (and #disconnect will still kick in anyway because you have the ensure block) :-)